### ISSUE TITLE
CompileNodeToValue: compile complex expressions

### ISSUE THREAD

**Issue by asgrim** [Roave/BetterReflection#64]
e.g. stuff like `["foo"."bar" => 2 * 3]`

They are constant expressions, but currently not supported.

**Comment by asgrim** [comment]
Also:

```php
<?php

class Foo
{
    const BAR = 123 * self::BAZ;
    const BAZ = 456 * self::BAR;
}

var_dump(Foo::BAR, Foo::BAZ);
```

**Comment by asgrim** [comment]
@ircmaxell I'd love it if you had any feedback on what the best way to implement this may be.

To summarise, given that we have an AST node (that represents a value):

```php
public function foo($x = null) {
```

Currently we have this [current CompileNodeToValue.php](https://github.com/Roave/BetterReflection/blob/336d9aa95e703907dd8e2f17ea8d3947796ed1f6/src/NodeCompiler/CompileNodeToValue.php)

Which takes the AST node (representing `null` in the example above), and evaluating `null` is pretty simple (it is a `ConstNode` so we can just return the value `null`).

However, it starts to get a little more complicated if you have an expression, e.g.:

```php
5 + (self::FOO * 3)
```

The AST tree needs to essentially be "compiled" now (beyond the rudimentary switch we have currently). The current proposal we have is in [proposed CompileNodeToValue.php](https://github.com/Roave/BetterReflection/blob/a79162c1dd1079e6ac468d0a25e601c47e65aa68/src/NodeCompiler/CompileNodeToValue.php) (part of PR #67), but basically uses the `PhpParser` printer to render PHP code (for whitelisted nodes) and then `eval()`'s the statement...

The main issues at the moment with this approach is:

* `ClassConstFetch` and `ConstFetch` (that isn't `true|false|null`) are difficult to evaluate - contextual information is required to do this. We've discussed this in issue #19 and feel that the best solution at the moment is to only resolve `ClassConstFetch`, otherwise there's a big rabbit hole. Bear in mind we could potentially be very limited to the scope we are able to look for defined constants for example (e.g. if we only have the source code of a single file, but the constant defined in a different file to that which we are reflecting from, then we would not be able to evaluate the constant).
* This code in #67 uses `eval()`... despite using a white list, I still feel uncomfortable using `eval()` to do this, although I'm not sure if there are any other approaches...

If you have any moments spare, I'd love it if we could pick your brains on this one! Thanks so much :)

**Comment by asgrim** [comment]
Closed in #95
